Add dropdown menus grouping similar tasks to sidebar

jadalilleboe opened the pull request:
![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/82007190/158223501-2bbfa640-0925-453c-b799-1cec4786db63.gif)

stefanv commented:
@jadalilleboe The foldable admin tabs is very nice, thank you! It isn't clear to me what the "Add" functionality does, or how you move items from the sidebar into the "Add" list.

stefanv commented:
Ah, see I thought "Add" was meant as a verb, and then I saw the `+` which re-enforced that that was a customizable set of tabs. What about `Other`?

Also, it looks like Telescopes appears under both admin and `Add` right now?

jadalilleboe commented:
updated!
![ezgif com-gif-maker (3)](https://user-images.githubusercontent.com/82007190/158416749-548a7cfb-b38c-4d9c-af16-57120e0cf056.gif)

stefanv commented:
This looks good; thanks @jadalilleboe 

@acrellin Do you remember why e94afeeba85f22edc5b5da6ca526bac7062ac6ff was necessary? It seems like that causes a lot of refetches every time a sidebar link gets clicked, which defies the purpose of the redux app cache.

acrellin commented:
Hey @stefanv, that was motivated by this ticket:
https://github.com/skyportal/skyportal/issues/1891
It appears we only rehydrate when the dashboard link is clicked.

On Tue, Mar 15, 2022 at 3:42 PM Stefan van der Walt <
***@***.***> wrote:

> This looks good; thanks @jadalilleboe <https://github.com/jadalilleboe>
>
> @acrellin <https://github.com/acrellin> Do you remember why e94afee
> <https://github.com/skyportal/skyportal/commit/e94afeeba85f22edc5b5da6ca526bac7062ac6ff>
> was necessary? It seems like that causes a lot of refetches every time a
> sidebar link gets clicked, which defies the purpose of the redux app cache.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/2757#issuecomment-1068535985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTPYIDRP5SLX4F3PUGDVAEG4TANCNFSM5QWFL7EQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Ari Crellin-Quick
pronouns: he/him/his (https://www.mypronouns.org/)

stefanv commented:
Thanks @acrellin! OK, so @jadalilleboe looks like the desired behavior is to only rehydrate when links are clicked that take you back to the frontpage.

jadalilleboe commented:
@stefanv ok, removed that!

stefanv commented:
Last comment, then we can put this in: can you do the spacing in `rem` instead of pixels?

See https://skyportal.io/docs/styling.html#sizes

jadalilleboe commented:
@stefanv fixed! sorry about that

stefanv commented:
@jadalilleboe What are your thoughts about the `sidebar` config now?

Should we modify the `sidebar` config to have sections (`main`, `other`, `admin`) with subitems?  Otherwise, this PR hard-codes some of the sidebar, which previously was fully customizable.

jadalilleboe commented:
@stefanv What do you mean it was fully customizable before? The only hard-coded subitems are user management and db stats, which were [previously also hard-coded](https://github.com/skyportal/skyportal/blob/master/static/js/components/SidebarAndHeader.jsx.template). I think this is because they're the only two subitems that need special checks for their visibility. 
Adding an item to either the main sidebar or the "other" group should still be customizable.

I guess if there are more subitems added in the future that need special checks, it would be good to have separate groupings for them in the config. But for now, since there's only one subitem in each "special" grouping (db stats is admin only and user management is for admins and user managers), I think it would be fine to leave them as hard-coded? What do you think?

stefanv commented:
What I suggest is that we split the list into three, so we have something like:

```
sidebar:
  main:
    - one
    - two
    - three
  other:
    - four
    - five
    - six
  admin:
    - only
    - admin
    - items
```

`main` and `admin` has special meaning. You can then choose whether you only want to allow `other`, or whether you also want any additional category, such as `fritz`---so users can build custom foldable menus (would be quite nice!).

Pseudocode:

```
for menu in sidebar:
    if menu == 'main':
        add_items_to_menu(root, sidebar['menu'])
    else:
        make_menu(menu, display_only_for_admin=(menu == 'admin'))
        add_items_to_menu(menu, sidebar['menu'], display_only_for_admin=(menu == 'admin'))
```

jadalilleboe commented:
@stefanv Oh ok, so there would be some sort of front end interaction for creating a custom foldable side menu and then this would get saved in the user's preferences? And the default would be whatever predefined subitems are in main, other, and admin

stefanv commented:
For now, no front-end interaction, just rendering the menu from the config file.

stefanv commented:
Yes, very nice!  You can make the permissions a list, and then that check becomes entirely generic (I like it!) and you don't need to check against any pre-existing values.

jadalilleboe commented:
updated version: removed icon on menus because it was hard to get lined up with the ones on the main sidebar. Long menu names are truncated.

https://user-images.githubusercontent.com/82007190/159752270-caf2b83f-bfb5-463a-bef2-cbfba7fd8909.mp4

stefanv reviewed (commented):
Very nicely done, @jadalilleboe.

stefanv commented:
Thank you very much!
